### PR TITLE
Conformance 1.2 Test Pack

### PR DESCRIPTION
**Pull request by robmoffat**

### CODE REVIEW

**Comment by robmoffat** [comment]
@kriswest I've moved all our 1.2 tests into a PR so that we can discuss on Wednesday and work towards a sign-off.  

I have a feeling you told me that there were some inaccuracies in these definitions, so please take a look and let me know in the review if anything needs changing.

thanks!

**Review by kriswest — CHANGES_REQUESTED** [review]
> I have a feeling you told me that there were some inaccuracies in these definitions, so please take a look and let me know in the review if anything needs changing.

I definitely did - both in the original issue and when the last couple of times we discussed getting sign-off on the cases and implementation from the FDC3 maintainers. The App Channel tests are nearly all wrong and not what I provided - during transcription into the github issue they turned into a copy of the user channel tests and are hence completely wrong (for example you do NOT join an app channel). There are a few others that are not fully defined, e.g.:

> `TargetedResolve2,3,4` Use the other ways of addressing apps (via ID, metadata) as described at the start of #18

Further, as discussed there is no way to relate the IDs in these cases to those in the implemented tests. Without that mapping I believe documentation of the cases has little value as they can only be compared to the test framework laboriously.

I cannot commit to correcting these cases for you - particularly if you want me to complete the 2.0 cases this week.

For the App channel cases specifically, look at the 2.0 test cases as I think the only change on those tests between 1.2 and 2.0 should be that `addContextListener`  was synchronous in 1.2 and went async in 2.0 (so example API calls have added `await`). Tests highlighted as different from 1.2 are (I think) just missing test cases from 1.2 not actual API differences.

**Comment by robmoffat** [comment]
> I definitely did - both in the original issue a

ok, I'll check that out, thanks.

> Further, as discussed there is no way to relate the IDs in these cases to those in the implemented tests. 

I'm going to resolve that problem in the conformance suite

> I cannot commit to correcting these cases for you - particularly if you want me to complete the 2.0 cases this week.

Please concentrate on finishing the 2.0 definitions. SL are ploughing through the ones you provided already now.

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:74]:
```suggestion
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | addContextListener |Call `testChannel.addContextListener(null, handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object  |
| B   | createChannel        | `const testChannel = fdc3. getOrCreateChannel("test-channel")`   |
| B   | Broadcast          | `testChannel.broadcast(<some instrument>)`   |
| A   | Receive Context    | Instrument object matches the one broadcast in 2 above.      |

-  `ACBasicUsage1` Perform above test 

| App | Step               | Details                                                                                                                                                        |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| B   | createChannel        | `fdc3. getOrCreateChannel("test-channel")`   |
| B   | Broadcast          | `testChannel.broadcast(<some instrument>)`   |
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | getCurrentContext |Call `testChannel.getCurrentContext()`<br>Check that the Context object returned is of the expected type.  |

-  `ACBasicUsage2` Perform above test.
-  `ACBasicUsage3` Perform above test, but in the final step call `getCurrentContext with a type name. I.e.  `testChannel.getCurrentContext("fdc3.instrument")`.
```

Comment (kriswest) [reply toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:74]: This is actually whats implemented for ACBasicUsage2 and ACBasicUsage3 - but I think they just duplicate ACContextHistoryTyped etc..

Hence, they should probably be removed (from both here and the implementation) after checking that they really are duplicates. @robmoffat

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:91]:
```suggestion
| App | Step               | Details                                                                                                                                                              |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | addContextListener | Call `testChannel.addContextListener("fdc3.instrument", handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object |
| B   | createChannel        | `const testChannel = await fdc3. getOrCreateChannel("test-channel")`   |
| B   | Broadcast          | `testChannel.broadcast()` the instrument context.<br>`testChannel.broadcast()` a contact context. |
| A   | Receive Context    | Instrument object matches the one broadcast in 2 above.<br>Check that the contact is not received.                                                                   |

-  `ACFilteredContext1`: Perform above test 
-  `ACFilteredContext2`: Perform above test, but add listeners for both `fdc3.instrument` and `fdc3.contact` in `addContextListener` step.  Both should be received. 
-  `ACFilteredContext3`: Perform above test, except creating a _different_ channel. Check that you _don't_ receive anything.
-  `ACUnsubscribe`: Perform above test, except that after creating the channel **A** then `unsubscribe()`s the listener it added to the channel. Check that **A** _doesn't_ receive anything.
-  `ACFilteredContext4`: Perform above test, except that after creating the channel **A** creates another channel with a further _different_ channel id.  Check that **A** _doesn't_ receive anything.
```

FYI ACFilteredContext5 is not possible (can't join/leave an app channel)

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:100]:
```suggestion
| A   | Receive Context    | `testChannel.getCurrentContext('fdc3.instrument')` returns the last instrument<br>testChannel.getCurrentContext('fdc3.contact')` returns the last broadcast contact                                                              |
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:98]:
```suggestion
| B   | createChannel        | `fdc3.getOrCreateChannel("test-channel")`   |
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:103]: n.b. implementation of this one is only testing the instrument type - it should be updated to do both types as written here @robmoffat

Comment (kriswest) [reply toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:103]: @Joe-Dunleavy FYI

**Review by kriswest — CHANGES_REQUESTED** [review]
took a quick look at the AppChannel tests and suggested some corrections. Not reviewed the rest in full again yet.

**Review by nkolba — COMMENTED** [review]
RE: Specific Context test for the `Open` API.  Is this saying that the Promise for Open should not resolve until the Context is received by they opened app (and rejects if Context is malformed)?   This behavior is not in the spec or API adds a lot more complexity to implementing this API and it is not clear what value it has for interop.   Implementations I'm familiar with resolve Open after the target application is launched.

**Comment by kriswest** [comment]
> RE: Specific Context test for the `Open` API. Is this saying that the Promise for Open should not resolve until the Context is received by they opened app (and rejects if Context is malformed)? This behavior is not in the spec or API adds a lot more complexity to implementing this API and it is not clear what value it has for interop. Implementations I'm familiar with resolve Open after the target application is launched.

@nkolba `AOpensBWithSpecificContext` seems fine to me - although the test definition is not actually stating the expected outcome all that well.

However, I am also confused by the tests in the next section (entitled 'Specific Context'), which is what I think you are referring to. In the first test, why is the `AppTimeout` error is thrown (as the context will be received - a filtered listener was not specified in the test description @robmoffat ).  However, it will fail the delivery (as expected) in `AOpensBNoListen`. In the final variation (`AOpensBMalformedContext`) the outcome seems wrong also - an invalid argument should probably cause an error. However, this is not specified anywhere I guess... so this case could go.
As to your point about whether the errors are specified in 1.2 - it's certainly poorly described:

>If opening errors, it returns an Error with a string from the [OpenError](https://fdc3.finos.org/docs/1.2/api/ref/Errors#openerror) enumeration.

but I think it is there. What is the purpose of the `OpenError.AppTimeout` error if not to be returned in that case. Likewise, it's not entirely clear what `OpenError.ErrorOnLaunch` is for.

If we can't agree on this one, dropping it is not the worst outcome. However, this test is required and the behavior more clearly specified in 2.0 under the OpenError docs: https://fdc3.finos.org/docs/api/ref/Errors#openerror

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (Joe-Dunleavy) [reply toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:74]: @robmoffat can confirm `ACContextHistoryTyped` is a duplicate of `ACBasicUsage3`. The only difference is that ACBasicUsage3 is checking context is received for a single fdc3.instrument type, whereas ACContextHistoryTyped checks for two context types. 

Let me know which one you want to remove and I'll remove it from the code

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (Joe-Dunleavy) [reply toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:102]: I've fixed this in code. Will create a PR once the comment I left above is resolved

**Comment by Joe-Dunleavy** [comment]
I've created a pull request to address the changes and issues mentioned here:

https://github.com/finos/FDC3-conformance-framework/pull/108

**Review by Joe-Dunleavy — APPROVED** [review]

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (greyseer256) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:69]:
```suggestion
| B   | createChannel        | `const testChannel = fdc3.getOrCreateChannel("test-channel")`   |
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (greyseer256) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:90]:
```suggestion
| B   | createChannel        | `const testChannel = await fdc3.getOrCreateChannel("test-channel")`   |
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (greyseer256) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:107]:
```suggestion
| A   | Receive Context    | `testChannel.getCurrentContext('fdc3.instrument')` returns the last broadcast instrument<br>`testChannel.getCurrentContext('fdc3.contact')` returns the last broadcast contact                                                              |
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (greyseer256) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:146]:
```suggestion
-  `AOpensBNoListen`: Skip `fdc3.addContextListener()` above. 
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (greyseer256) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:147]:
```suggestion
-  `AOpensBMultipleListen`:  **B** performs `fdc3.addContextListener('fdc3.instrument') prior to the existing `addContextListener`.  The correct context listener should receive the context, and the promise resolves successfully
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (greyseer256) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:171]:
```suggestion
-  `IntentAppDMultiple2`: Calls `fdc3.findIntent(‘sharedTestingIntent1’, 'testContextX')`.  Receives promise containing an appIntent with metadata containing `sharedTestingIntent` and only **A** and **B** app metadata.
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (greyseer256) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:172]:
```suggestion
-  `IntentAppDMultiple3`: Calls `fdc3.findIntent(‘sharedTestingIntent1’, 'testContextY')`.  Receives promise containing an appIntent with metadata containing `sharedTestingIntent` and only **B** app metadata.
```

**Comment by robmoffat** [comment]
@nkolba please review

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:50]: this should replicate the next step (so that you have a channel to join)

```suggestion
| A   | 2. joinChannel        | `fdc3.getSystemChannels()`<br>Check channels are returned.<br>Call `fdc3.joinChannel()` on first non-global channel. Check that there is an unsubscribe function on the returned object                                                                                                                                                                                                                                                                                |
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:96]: Small clarification:
```suggestion
-  `ACFilteredContext3`: Perform above test, except creating a _different_ channel in app B. Check that you _don't_ receive anything (as the channels don't match).
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:98]: This is wrong - if you create a further channel and still have listeners on the other channel it should still receive stuff - this test is supposed to check that the additional channel does NOT break the first.

```suggestion
-  `ACFilteredContext4`: Perform above test, except that after creating the channel **A** creates another channel with a further _different_ channel id and adds a further context listener to it.  Check that **A** is still able to receive context on the first channel (i.e. it is unaffected by the additional channel) and does *NOT* receive anything on the second channel.
```

@Joe-Dunleavy and @robmoffat please check that this is how the test was implemented.

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:133]: The other tests in this group have 3 variations. This doesn't appear to (and probably doesn't need to) - so which variation should it be using (doesn't matter, but be specific)?

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:148]: Describe how its malformed - e.g. no type field.

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:147]:
```suggestion
-  `AOpensBMultipleListen`:  **B** performs `fdc3.addContextListener('fdc3.instrument') prior to the existing `addContextListener`.  Check that the specific context listener for `fdc3.instrument` does not receive the context, but that the listener with an unspecified type does, and the promise resolves successfully.
```

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:190]: I'd prefer variations to be spelled out here rather than referencing other tests or just 'and variations' - but this is otherwise correct.

**Review by kriswest — CHANGES_REQUESTED** [review]
One error found, a few other minor suggestions - its very close!

**Review by greyseer256 — APPROVED** [review]

**Review by ggeorgievx — APPROVED** [review]

**Review by mistryvinay — APPROVED** [review]
Great effort on this team. Looks good.

**Review by nkolba — APPROVED** [review]

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:15]: @robmoffat Kalina's PR is adding a BasicJC2 and BasicJC3 - BasicJC2 makes sense and needs adding here, but we can't do BasicJC3 (for the same reason others don't want to test OpenError.AppTimeout, it isn't clear that it could/should be returned if the channel was not in system channels). In fact *most* implementations of 1.2 allow you to join AppChannels - it was only in 2.0 that we started to clarify the intention that it should only be user channels that you join (and documented the relevant error).

File: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md
Comment (openfin-johans) [inline toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md:69]: Let's remove the space after fdc3.

**Review by openfin-johans — CHANGES_REQUESTED** [review]
Looks good as a first version - found a small typo

**Review by openfin-johans — APPROVED** [review]
LGTM

**Review by kriswest — APPROVED** [review]
LGTM

**Pull request merged**
